### PR TITLE
Community and Collection Tree

### PR DESCRIPTION
**Pull request by MarieVerdonck**
### Overview
This PR adds a component to render the community-collection tree. It loads only what you see, but is expandable to any requested level through expanding the (top) community nodes and loading more results via pagination. The structure itself is composed of a Material CDK tree.

### Material CDK
The trees makes use of the [Material CDK library](https://material.angular.io/cdk), which is described by Google as follows:

>The Component Dev Kit (CDK) is a set of tools that implement common interaction patterns whilst being unopinionated about their presentation. It represents an abstraction of the core functionalities found in the Angular Material library, without any styling specific to Material Design. Think of the CDK as a blank state of well-tested functionality upon which you can develop your own bespoke components.

Material CDK is compatible with Bootstrap and doesn’t use the material css framework. In an upcoming PR, we also intend to use the drag & drop module in material CDK for reordering bitstreams on the edit item page.

### Implementation Details
CDK-tree provides a tree structure based on a list of nodes defined via its dataSource attribute. In this case we use the flat tree, so the dataSource is a list of flat nodes which each contain a level attribute, indicating how far into the tree it should be rendered. Initially the list of flat nodes is comprised of just a page-limited set of top communities, which is expandable to more top communities via show more, as well as expandable deeper into the tree via the chevron. 

The creation of the flat node tree is recursive so it can handle levels of communities and subcommunities higher than 2. All ‘children’ of a community (its subcommunities and collections) are also paginated and can be expanded (via show more nodes).

The rendering of the nodes is split into expandable nodes (with children), non-expandable (childless) nodes and show more nodes, so that their appearance and behaviour may be changed accordingly. If the tree is loading/gets expanded (either through chevron or show more functionality) a loading indication is shown on the corresponding location in the tree.

The description of the communities and collections are truncated to 3 lines; the nr of top communities per page is set to 10 and the number of sub communities and collections per page to 3, hardcoded in the community-list service.

### Testing & Access
This communities & collections tree can be accessed via /community-list path or via the menu: All of DSpace > Communities & Collections.

Test spec files for the component and the service which creates the list of flat nodes required were also created.

### CODE REVIEW

**Comment by mspalti** [comment]
Hi @MarieVerdonck!

This PR initially worked for me as you describe. I did run into problems over time, and as I went farther down into the tree.  Here's what I saw when testing against the demo server:

When I expanded the MASENO UNIVERSITY node, the loading indicator appeared, then two sub-communities.  The first sub-community, SCIENCE AND INNOVATION, includes the right-facing chevron, but when I click it, no new information is provided. SCIENCE AND INNOVATION is an empty collection with no items or sub-communities, so I think the chevron should not appear in this case? I ran into several other examples like this. 

Also, I had to attend a meeting, and when I returned I clicked on a chevron (with causes the list to be re-rendered) and saw that the list was completely empty. I also noticed that when browsing between the list and sub-items, eventually after enough time earlier pages disappear from the list. These seem like caching issues.

One other thing I noticed is that when using the browser navigation to return the the list after viewing a community/collection the expanded nodes are collapsed. For me, this seems like something we should eventually fix. Nodes that have been expanded should remain expanded during user navigation.

I'll look again tomorrow and let you know if I find more to report.

File: src/app/community-list-page/community-list-datasource.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { CommunityListService, FlatNode } from './community-list-service';
+import { CollectionViewer, DataSource } from '@angular/cdk/typings/collections';
+import { BehaviorSubject, Observable, } from 'rxjs';
+import { finalize, take, } from 'rxjs/operators';
+
+export class CommunityListDatasource implements DataSource<FlatNode> {
```
Comment (tdonohue) [inline src/app/community-list-page/community-list-datasource.ts]: Please add TypeDocs/comments to describe this new Class

File: src/app/community-list-page/community-list-page.component.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { Component } from '@angular/core';
+
+@Component({
```
Comment (tdonohue) [inline src/app/community-list-page/community-list-page.component.ts]: Please add TypeDocs/comments to describe this Component.  All new classes should have basic TypeDocs please.

File: src/app/community-list-page/community-list-page.module.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { CommonModule } from '@angular/common';
+import { NgModule } from '@angular/core';
+import { SharedModule } from '../shared/shared.module';
+import { CommunityListPageComponent } from './community-list-page.component';
+import { CommunityListPageRoutingModule } from './community-list-page.routing.module';
+import { CommunityListComponent } from './community-list/community-list.component';
+import { CdkTreeModule } from '@angular/cdk/tree';
+
+@NgModule({
```
Comment (tdonohue) [inline src/app/community-list-page/community-list-page.module.ts]: Same here, a basic description of the Module in TypeDocs is needed

File: src/app/community-list-page/community-list-page.routing.module.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { NgModule } from '@angular/core';
+import { RouterModule } from '@angular/router';
+import { CdkTreeModule } from '@angular/cdk/tree';
+
+import { CommunityListPageComponent } from './community-list-page.component';
+import { CommunityListService } from './community-list-service';
+
+@NgModule({
```
Comment (tdonohue) [inline src/app/community-list-page/community-list-page.routing.module.ts]: Here too, Typedocs please

File: src/app/community-list-page/community-list-service.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    level: level,
+    isExpanded: false,
+    parent: parent,
+    payload: new ShowMoreFlatNode(),
+    isShowMoreNode: true,
+});
+
+// tslint:disable-next-line:max-classes-per-file
+@Injectable()
+export class CommunityListService {
```
Comment (tdonohue) [inline src/app/community-list-page/community-list-service.ts]: Again, TypeDocs please.  I'd also appreciate comments to describe the various exported `const` above.

File: src/app/community-list-page/community-list/community-list.component.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
+import { Component, OnInit } from '@angular/core';
+import { CommunityListService, FlatNode } from '../community-list-service';
+import { CommunityListDatasource } from '../community-list-datasource';
+import { FlatTreeControl } from '@angular/cdk/tree';
+import { isEmpty } from '../../shared/empty.util';
+
+@Component({
```
Comment (tdonohue) [inline src/app/community-list-page/community-list/community-list.component.ts]: TypeDocs please at the Class level (most of the methods already have comments, and that's great to see)

**Comment by mspalti** [comment]
@MarieVerdonck : The errors that @tdonohue noticed appear to be the result of forceBypassCache being removed from `DataService`. Merging the new code from master should fix the problem.
https://github.com/DSpace/dspace-angular/pull/468

**Comment by MarieVerdonck** [comment]
Which nodes were previously expanded are now kept in the store, so the tree state can persist across the application.
AoT compilation errors were indeed because of tests that needed to be altered because of the changes made to fix the incorrect chevron rendering.

**Comment by mspalti** [comment]
Thanks.  AoT compilation is working for me.

File: src/app/community-list-page/community-list-service.ts [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+   * Gets all top communities, limited by page, and transforms this in a list of flatNodes.
+   * @param expandedNodes     List of expanded nodes; if a node is not expanded its subCommunities and collections need
+   *                            not be added to the list
+   */
+  loadCommunities(expandedNodes: FlatNode[]): Observable<FlatNode[]> {
+    const res = this.payloads$.map((payload) => {
+      return payload.pipe(
+        take(1),
+        switchMap((result: PaginatedList<Community>) => {
+          return this.transformListOfCommunities(result, 0, null, expandedNodes);
```
[the remark this answers is not in the document]
Comment (mspalti) [reply src/app/community-list-page/community-list-service.ts]: The same blank screen appears if I wait long enough before trying to expand a community in the list.  It seems like this happens whenever the `payload$` observable in CommunityListService has not been replenished after a cache timeout. Again, I'm a bit puzzled.

**Comment by mspalti** [comment]
When you expand the Kaimosi Friends University College community there's something odd happening. There's a "show more" button under "School of Education and Social Sciences" and it's not clear why. The button does reveal another collection, but that's unexpected since it is followed by additional communities and a second "show more" button.

**Comment by MarieVerdonck** [comment]
> When you expand the Kaimosi Friends University College community there's something odd happening. There's a "show more" button under "School of Education and Social Sciences" and it's not clear why. The button does reveal another collection, but that's unexpected since it is followed by additional communities and a second "show more" button.

There is a show more button to show more subcommunities and one to show more collections there. In this case the first 'show more' (for the  subcommunities) indicates that there is another subcommunity, namely 'School of Science'. But since this is a subcommunity without its own subcommunities or collections it does not have a chevron and appears to be a collection, which caused the confusion I'm guessing. It is possible to change the appearance of collections/communities so they are not the same, but this would subtract from the uniformity of the tree possibly.

**Comment by artlowel** [comment]
@mspalti The caching issue is unrelated to this PR. There is a ticket for it: #216

**Comment by mspalti** [comment]
Thanks for explaining the layout @MarieVerdonck , I think something should be done to make the distinction more obvious, but that's a design detail that can probably wait. (Perhaps the question merits a separate issue or perhaps it's part of a separate design review?) 

Since caching is also a known issue I can approve the PR. Thanks for your work on this!

**Comment by tdonohue** [comment]
@MarieVerdonck : this currently has build failures in Travis CI

**Comment by tdonohue** [comment]
Per today's meeting discussion, we recommend that the removed `e2e` test be added in a separate small PR so that we can further analyze it as to why Travis does not succeed.  

@artlowel  noted in today's meeting that this `e2e` test works fine locally & via Docker -- it's only in Travis where failures arise.  This implies that it might be specific to how Travis is running e2e tests....so creating a separate PR would let us more closely analyze how to potentially fix it.

**Comment by artlowel** [comment]
The new PR for the test can be found here: #538

**Pull request merged**